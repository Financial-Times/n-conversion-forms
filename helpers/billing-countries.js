import { billingCountries } from 'n-common-static-data';

export { billingCountries };
