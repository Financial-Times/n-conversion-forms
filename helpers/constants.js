const printRegions = {
	cemeaV1: 'CEMEA_V1',
	cemeaV2: 'CEMEA_V2',
	apac: 'APAC',
};

const addressLine1AutoComplete = 'address-line1';
const addressLine2AutoComplete = 'address-line2';
const addressLine3AutoComplete = 'address-line3';

module.exports = {
	printRegions,
	addressLine1AutoComplete,
	addressLine2AutoComplete,
	addressLine3AutoComplete,
};
