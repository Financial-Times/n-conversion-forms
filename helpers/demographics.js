const { demographics } = require('n-common-static-data');

export { demographics };
