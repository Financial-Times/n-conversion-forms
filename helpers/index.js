module.exports = {
	'ncf-countries': require('./ncf-countries'),
	'ncf-common-data': require('./ncf-common-data'),
	cemeaV1ISO: require('./supportedCountries').cemeaV1ISO,
	cemeaV2ISO: require('./supportedCountries').cemeaV2ISO,
	apacISO: require('./supportedCountries').apacISO,
	identifyFTShippingZone: require('./supportedCountries')
		.identifyFTShippingZone,
	countriesSupported: require('./supportedCountries').countriesSupported,
	countriesSupportedISO: require('./supportedCountries').countriesSupportedISO,
	supportedPostcodeValidators: require('./supportedPostcodeValidators')
		.supportedPostcodeValidators,
	allSupportedPostcodeExamples: require('./supportedPostcodeExamples')
		.allSupportedPostcodeExamples,
	demographics: require('./demographics').demographics,
	billingCountries: require('./billing-countries').billingCountries,
};
