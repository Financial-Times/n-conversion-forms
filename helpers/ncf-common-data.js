const _get = require('lodash.get');
const COMMON_DATA = require('n-common-static-data');

module.exports = function ({ hash, fn }) {
	const context = Object.assign({}, { [hash.export]: _get(COMMON_DATA, hash.import) }, this);
	return fn(context);
};
