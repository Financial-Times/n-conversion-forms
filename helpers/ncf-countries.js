const { countries } = require('n-common-static-data').billingCountries;

const MINIMUM_TO_GROUP = 50;
const MINIMUM_TO_SHOW_GROUPS = 2;

module.exports = function ({ hash = {}, fn }) {
	const filter = Array.isArray(hash.filterList) ? hash.filterList : false;
	const value = hash.value;

	let data = countries;

	// Only show countries in the filter
	if (filter) {
		data = data.filter(item => filter.includes(item.code));
	}

	// Select the country
	if (value) {
		data = data.map(item => {
			item.selected = item.code === value;
			return item;
		});
	}

	// Group countries
	if (data.length >= MINIMUM_TO_GROUP) {
		data = groupCountries(data);
	}

	const context = Object.assign({ countries: data }, this);
	return fn(context);
};

/**
 * Produce a frequently used option group
 * @param {Array} countries
 * @return {Array}
 */
function groupCountries (countries) {
	const frequentlyUsed = ['GBR', 'USA', 'JPN', 'FRA', 'CAN'];

	const frequentlyUsedCountries = countries
		.filter(item => frequentlyUsed.includes(item.code))
		.map(item => Object.assign({}, item))
		.sort((a, b) => frequentlyUsed.indexOf(a.code) - frequentlyUsed.indexOf(b.code));

	// Not worth showing return standard countries list
	if (frequentlyUsedCountries.length < MINIMUM_TO_SHOW_GROUPS) {
		return countries;
	}

	// If selected country is in frequently used only leave that one selected
	// otherwise the user is shown the lowest selected option
	const selected = frequentlyUsedCountries.find(item => item.selected);
	if (selected) {
		countries.forEach(item => item.selected = false);
	}

	return [
		{ label: 'Frequently Used', countries: frequentlyUsedCountries },
		{ label: 'Alphabetical', countries }
	];
}
