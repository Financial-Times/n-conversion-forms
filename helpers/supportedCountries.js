const supportedCountriesMasterList = {
	cemeaV1: {
		AUT: 'Austria',
		BEL: 'Belgium',
		BGR: 'Bulgaria',
		HRV: 'Croatia',
		CZE: 'Czechia',
		DNK: 'Denmark',
		FIN: 'Finland',
		DEU: 'Germany',
		GRC: 'Greece',
		HUN: 'Hungary',
		ISL: 'Iceland',
		ITA: 'Italy',
		LIE: 'Liechtenstein',
		LTU: 'Lithuania',
		NLD: 'Netherlands',
		NOR: 'Norway',
		POL: 'Poland',
		PRT: 'Portugal',
		QAT: 'Qatar',
		ROU: 'Romania',
		RUS: 'Russian Federation',
		SVK: 'Slovakia',
		SVN: 'Slovenia',
		ZAF: 'South Africa',
		ESP: 'Spain',
		SWE: 'Sweden',
		CHE: 'Switzerland'
	},
	cemeaV2: { FRA: 'France', LUX: 'Luxembourg', MCO: 'Monaco' },
	apac: {
		CHN: 'China',
		HKG: 'Hong Kong',
		IND: 'India',
		IDN: 'Indonesia',
		JPN: 'Japan',
		KOR: 'Korea, Republic of',
		MYS: 'Malaysia',
		PHL: 'Philippines',
		SGP: 'Singapore',
		TWN: 'Taiwan, Province of China',
		THA: 'Thailand'
	},
	GBR: 'United Kingdom',
	USA: 'United States',
	CAN: 'Canada'
};


const cemeaV1ISO = Object.keys(supportedCountriesMasterList.cemeaV1);
const cemeaV2ISO = Object.keys(supportedCountriesMasterList.cemeaV2);
const apacISO = Object.keys(supportedCountriesMasterList.apac);
const countriesSupported = Object.assign({}, ...(function _flatten (o) { return [].concat(...Object.keys(o).map(k => typeof o[k] === 'object' ? _flatten(o[k]) : ({[k]: o[k]})));}(supportedCountriesMasterList)));


module.exports = { cemeaV1ISO, cemeaV2ISO, apacISO, countriesSupported };
