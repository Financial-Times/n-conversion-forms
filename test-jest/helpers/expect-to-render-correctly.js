import ReactDOMServer from 'react-dom/server';

export const expectToRenderCorrectly = {
	toRenderCorrectly(reactComponent, props) {
		// Render the React component with props.
		const ReactComponentRendered = ReactDOMServer.renderToStaticMarkup(
			reactComponent(props)
		);

		// eslint-disable-next-line jest/no-standalone-expect
		expect(ReactComponentRendered).toMatchSnapshot();

		// This point is reached when the above assertions were successful.
		// The test should therefore always pass, that means it needs to be
		// `true` when used normally, and `false` when `.not` was used.
		return {
			pass: !this.isNot,
		};
	},
};
