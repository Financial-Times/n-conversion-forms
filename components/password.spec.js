import { Password } from './index';
import { expectToRenderCorrectly } from '../test-jest/helpers/expect-to-render-correctly';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

expect.extend(expectToRenderCorrectly);

describe('Password', () => {
	it('render a password input with a label where input ID and Name are the same', () => {
		const props = {};

		expect(Password).toRenderCorrectly(props);

		const renderedPassword = mount(Password(props));
		const inputElement = renderedPassword.find('#password');
		expect(inputElement.prop('name')).toBe('password');
	});

	it('can have a different label', () => {
		const props = {
			label: 'Current password',
			inputId: 'passwordWithCustomLabel',
		};

		const renderedPassword = mount(Password(props));
		const inputElement = renderedPassword.find(
			`label[htmlFor="${props.inputId}"] .o-forms-title__main`
		);
		expect(inputElement.text()).toBe(props.label);
	});

	it('can have different placeholder text', () => {
		const props = {
			placeholder: 'Please enter your current password',
			inputId: 'passwordWithCustomPlaceholder',
		};

		const renderedPassword = mount(Password(props));
		const inputElement = renderedPassword.find(`#${props.inputId}`);
		expect(inputElement.prop('placeholder')).toBe(props.placeholder);
	});

	it('can have an input ID different from an input name', () => {
		const props = {
			inputId: 'i-m-kebab-case',
			inputName: 'ImPascalCase',
		};

		const renderedPassword = mount(Password(props));
		const inputElement = renderedPassword.find(`#${props.inputId}`);
		expect(inputElement.prop('name')).toBe(props.inputName);
	});

	it('can render a disable input', () => {
		const props = {
			isDisabled: true,
		};

		expect(Password).toRenderCorrectly(props);
	});

	it('can render a pattern attribute', () => {
		const props = {
			pattern: 'Whatever',
		};

		expect(Password).toRenderCorrectly(props);
	});

	it('can render as an Error', () => {
		const props = {
			hasError: true,
		};

		expect(Password).toRenderCorrectly(props);
	});

	it('can render as an Unknown user', () => {
		const props = {
			unknownUser: true,
		};

		expect(Password).toRenderCorrectly(props);
	});

	it('can have different description text', () => {
		const props = {
			description: 'Keep this a secret!',
			inputId: 'passwordWithCustomDescription',
		};

		const renderedPassword = mount(Password(props));
		const passwordDescription = renderedPassword.find('.o-forms-title__prompt');
		expect(passwordDescription.text()).toBe(props.description);
	});

	it('can render without a description', () => {
		const props = {
			showDescription: false,
			inputId: 'passwordWithoutDescription',
		};

		const renderedPassword = mount(Password(props));
		const inputElement = renderedPassword.find(`#${props.inputId}`);
		expect(renderedPassword.exists('.o-forms-title__prompt')).toBe(false);
		expect(inputElement.prop('aria-describedby')).toBeUndefined();
	});

	it('can render without the show password checkbox', () => {
		const props = {
			hasShowPassword: false,
			inputId: 'passwordWithoutShowPassword',
		};

		const renderedPassword = mount(Password(props));
		expect(
			renderedPassword.exists('input[data-trackable="field-show-password"]')
		).toBe(false);
	});
});
