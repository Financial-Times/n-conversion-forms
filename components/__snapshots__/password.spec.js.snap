// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password can render a disable input 1`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
           disabled
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password can render a disable input 2`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
           disabled
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password can render a pattern attribute 1`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
           pattern="Whatever"
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password can render a pattern attribute 2`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
           pattern="Whatever"
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password can render as an Error 1`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix o-forms-input--invalid">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password can render as an Error 2`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix o-forms-input--invalid">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password can render as an Unknown user 1`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field js-unknown-user-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password can render as an Unknown user 2`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field js-unknown-user-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password render a password input with a label where input ID and Name are the same 1`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;

exports[`Password render a password input with a label where input ID and Name are the same 2`] = `
<div id="passwordField"
     class="o-forms-field ncf__password-field"
     data-validate="required,password"
>
  <label for="password"
         class="o-forms-title"
  >
    <span class="o-forms-title__main">
      Password
    </span>
    <span class="o-forms-title__prompt">
      Use 8 or more characters with a mix of letters, numbers &amp; symbols
    </span>
  </label>
  <div class="o-forms-input o-forms-input--password o-forms-input--checkbox o-forms-input--suffix">
    <input type="password"
           id="password"
           name="password"
           placeholder="Enter a password"
           autocomplete="new-password"
           data-trackable="field-password"
           aria-required="true"
           required
    >
    <label class="ncf__password-field--show-password">
      <input type="checkbox"
             id="showPassword"
             name="showPassword"
             data-trackable="field-show-password"
             aria-label="Show password"
      >
      <span class="o-forms-input__label"
            aria-hidden="true"
      >
        Show password
      </span>
    </label>
    <div class="o-forms-input__error">
      Please enter a valid password
    </div>
  </div>
</div>
`;
